tests/node_test.cc

```diff
@@ -249,4 +250,51 @@ TEST(NodeTest, MultipleSubscriber) {
     }
 }
 
+TEST(NodeTest, StrandSubscriber) {
+    static constexpr std::size_t kThreadNum       = 4;
+    constexpr std::size_t        kNumOfSubChannel = 4;
+    constexpr std::size_t        kMessageNumber   = 50000;
+
+    using TestMessageType = TestMessage<11>;
+
+    CRNode publisher;
+    CRNode subscriber;
+
+    auto runner = JobRunner::MakeJobRunner(JobRunner::Config{
+        .thread_num_ = kThreadNum,
+    });
+    subscriber.SetRunner(runner);
+
+    std::mutex                                cv_mutex;
+    std::condition_variable                   cv;
+    std::array<std::size_t, kNumOfSubChannel> channel_counter{0};
+    std::array<bool, kNumOfSubChannel>        complete{0};
+
+    for (std::size_t channel_subid = 0; channel_subid < kNumOfSubChannel; ++channel_subid) {
+        subscriber.Subscribe<TestMessageType>(
+            channel_subid,
+            [&counter = channel_counter[channel_subid], &complete = complete[channel_subid], &cv_mutex, &cv](
+                const std::shared_ptr<TestMessageType>& message) {
+                EXPECT_EQ(counter++, message->value_);
+                if (counter == kMessageNumber) {
+                    std::unique_lock cv_lock(cv_mutex);
+                    complete = true;
+                    cv.notify_all();
+                }
+            },
+            /* allow_concurrency = */ false);
+    }
+
+    for (std::size_t msg_idx = 0; msg_idx < kMessageNumber; ++msg_idx) {
+        for (std::size_t channel_subid = 0; channel_subid < kNumOfSubChannel; ++channel_subid) {
+            publisher.Publish(channel_subid, std::make_shared<TestMessageType>(msg_idx));
+        }
+    }
+
+    std::unique_lock cv_lock(cv_mutex);
+    cv.wait(cv_lock, [&complete]() {
+        return std::reduce(complete.begin(), complete.end(), true, std::logical_and<void>());
```

Comment:
unseq

Reply:
It is a very simply task and unseq cannot bring any meaningful change.

Before unseq
```
~/cris-core$ for i in {1..10}; do bazel test --config=prof --test_output=all --nocache_test_results //tests:node_test 2>&1 | grep 'OK.*NodeTest.StrandSubscriber'; done
[       OK ] NodeTest.StrandSubscriber (1348 ms)
[       OK ] NodeTest.StrandSubscriber (456 ms)
[       OK ] NodeTest.StrandSubscriber (631 ms)
[       OK ] NodeTest.StrandSubscriber (392 ms)
[       OK ] NodeTest.StrandSubscriber (493 ms)
[       OK ] NodeTest.StrandSubscriber (1331 ms)
[       OK ] NodeTest.StrandSubscriber (527 ms)
[       OK ] NodeTest.StrandSubscriber (444 ms)
[       OK ] NodeTest.StrandSubscriber (443 ms)
[       OK ] NodeTest.StrandSubscriber (396 ms)
```
After unseq
```
~/cris-core$ for i in {1..10}; do bazel test --config=prof --test_output=all --nocache_test_results //tests:node_test 2>&1 | grep 'OK.*NodeTest.StrandSubscriber'; done
[       OK ] NodeTest.StrandSubscriber (392 ms)
[       OK ] NodeTest.StrandSubscriber (896 ms)
[       OK ] NodeTest.StrandSubscriber (432 ms)
[       OK ] NodeTest.StrandSubscriber (516 ms)
[       OK ] NodeTest.StrandSubscriber (620 ms)
[       OK ] NodeTest.StrandSubscriber (357 ms)
[       OK ] NodeTest.StrandSubscriber (392 ms)
[       OK ] NodeTest.StrandSubscriber (1378 ms)
[       OK ] NodeTest.StrandSubscriber (543 ms)
[       OK ] NodeTest.StrandSubscriber (436 ms)
```
CPU:
```
lscpu
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Byte Order:                      Little Endian
Address sizes:                   40 bits physical, 48 bits virtual
CPU(s):                          4
On-line CPU(s) list:             0-3
Thread(s) per core:              1
Core(s) per socket:              4
Socket(s):                       1
NUMA node(s):                    1
Vendor ID:                       GenuineIntel
CPU family:                      6
Model:                           63
Model name:                      DO-Regular
Stepping:                        2
CPU MHz:                         2494.136
BogoMIPS:                        4988.27
Virtualization:                  VT-x
Hypervisor vendor:               KVM
Virtualization type:             full
L1d cache:                       128 KiB
L1i cache:                       128 KiB
L2 cache:                        16 MiB
```

Reply:
1. You need gcc, IIRC clang doesn't have real parallel TS support.
2. If it doesn't help, either STL impl is bad or O3 is great, either way we're providing extra info, e.g. logic short-circuit or fp eval order doesn't matter.
3. Probably you should double check if bitwise is better, because short-circuit introduces an extra branch and most likely reduce doesn't have early termination to take advantage of that.

Reply:
Will try, but since this is not a very expensive parts of the test, we should not spend too much time on it.

Reply:
Retried with GCC-10, and I do not see much difference. This is a straightforward reduce operation, maybe it can be optimized very well by the compiler, or parallelism brings more overhead than saved. I don't feel it is a problem worth spending more time on.
```
chenhao@chenhao:~/cris-core$ grep "std::execution" tests/node_test.cc && for i in {1..10}; do bazel test --config=prof --test_output=all --nocache_test_results //tests:n
ode_test 2>&1 | grep 'OK.*NodeTest.StrandSubscriber'; done                           
        return std::reduce(std::execution::seq, complete.begin(), complete.end(), true, std::logical_and<void>());
[       OK ] NodeTest.StrandSubscriber (458 ms)
[       OK ] NodeTest.StrandSubscriber (429 ms)
[       OK ] NodeTest.StrandSubscriber (540 ms)
[       OK ] NodeTest.StrandSubscriber (465 ms)
[       OK ] NodeTest.StrandSubscriber (458 ms)
[       OK ] NodeTest.StrandSubscriber (476 ms)
[       OK ] NodeTest.StrandSubscriber (445 ms)
[       OK ] NodeTest.StrandSubscriber (1505 ms)
[       OK ] NodeTest.StrandSubscriber (351 ms)
[       OK ] NodeTest.StrandSubscriber (486 ms)
chenhao@chenhao:~/cris-core$ grep "std::execution" tests/node_test.cc && for i in {1..10}; do bazel test --config=prof --test_output=all --nocache_test_results //tests:n
ode_test 2>&1 | grep 'OK.*NodeTest.StrandSubscriber'; done
        return std::reduce(std::execution::par, complete.begin(), complete.end(), true, std::logical_and<void>());
[       OK ] NodeTest.StrandSubscriber (1293 ms)
[       OK ] NodeTest.StrandSubscriber (709 ms)
[       OK ] NodeTest.StrandSubscriber (353 ms)
[       OK ] NodeTest.StrandSubscriber (487 ms)
[       OK ] NodeTest.StrandSubscriber (468 ms)
[       OK ] NodeTest.StrandSubscriber (418 ms)
[       OK ] NodeTest.StrandSubscriber (1015 ms)
[       OK ] NodeTest.StrandSubscriber (1405 ms)
[       OK ] NodeTest.StrandSubscriber (479 ms)
[       OK ] NodeTest.StrandSubscriber (358 ms)
chenhao@chenhao:~/cris-core$ grep "std::execution" tests/node_test.cc && for i in {1..10}; do bazel test --config=prof --test_output=all --nocache_test_results //tests:node_test 2>&1 | grep 'OK.*NodeTest.StrandSubscriber'; done
        return std::reduce(std::execution::unseq, complete.begin(), complete.end(), true, std::logical_and<void>());
[       OK ] NodeTest.StrandSubscriber (386 ms)
[       OK ] NodeTest.StrandSubscriber (476 ms)
[       OK ] NodeTest.StrandSubscriber (430 ms)
[       OK ] NodeTest.StrandSubscriber (416 ms)
[       OK ] NodeTest.StrandSubscriber (468 ms)
[       OK ] NodeTest.StrandSubscriber (1159 ms)
[       OK ] NodeTest.StrandSubscriber (1555 ms)
[       OK ] NodeTest.StrandSubscriber (558 ms)
[       OK ] NodeTest.StrandSubscriber (1685 ms)
[       OK ] NodeTest.StrandSubscriber (653 ms)
chenhao@chenhao:~/cris-core$ grep "std::execution" tests/node_test.cc && for i in {1..10}; do bazel test --config=prof --test_output=all --nocache_test_results //tests:node_test 2>&1 | grep 'OK.*NodeTest.StrandSubscriber'; done
        return std::reduce(std::execution::par_unseq, complete.begin(), complete.end(), true, std::logical_and<void>());
[       OK ] NodeTest.StrandSubscriber (476 ms)
[       OK ] NodeTest.StrandSubscriber (623 ms)
[       OK ] NodeTest.StrandSubscriber (373 ms)
[       OK ] NodeTest.StrandSubscriber (648 ms)
[       OK ] NodeTest.StrandSubscriber (493 ms)
[       OK ] NodeTest.StrandSubscriber (445 ms)
[       OK ] NodeTest.StrandSubscriber (420 ms)
[       OK ] NodeTest.StrandSubscriber (400 ms)
[       OK ] NodeTest.StrandSubscriber (430 ms)
[       OK ] NodeTest.StrandSubscriber (456 ms)
```

Reply:
> or parallelism brings more overhead than saved.

unseq is vectorization

> Will try, but since this is not a very expensive part of the test, we should not spend too much time on it.

Agree we should not waste time on this, though it's still good to give compiler extra hint for future improvement on the compile side.